### ISSUE TITLE
Findbugs is reporting false positive bugs SA_LOCAL_SELF_COMPARISON when using instanceof pattern matching 

### ISSUE THREAD

**Issue by mpet** [spotbugs/sonar-findbugs#876]
## Issue Description

When running sonar rules on java code that has the following construct:

https://docs.oracle.com/en/java/javase/17/language/pattern-matching-instanceof-operator.html#GUID-843060B5-240C-4F47-A7B0-95C42E5B08A7

Sonarqube rule: findbugs:SA_LOCAL_SELF_COMPARISON

gives a false positive.

## Environment

| Component          | Version |
| ------------------ | ------- |
| SonarQube          | Community EditionVersion 9.9 (build 65466)  |
| Sonar-FindBugs     | FindBugs 4.2.3 / SpotBugs 4.7.3  |
| Maven              | 3.8.5  |
| Gradle             |    |
| Java               | 17   |

## Code (If needed)

```java
private static X509Certificate readCertificatePEMFile(String certificatePemFilePath)
            throws IOException, CertificateException {
        if (certificatePemFilePath != null) {
            File certificatePemFile = getCertificatePEMFile(certificatePemFilePath);
            if (certificatePemFile.exists() && certificatePemFile.canRead()) {
                try (InputStream inStream = new FileInputStream(certificatePemFile)) {
                    try (PEMParser pemParser = new PEMParser(new InputStreamReader(inStream))) {
                        Object object = pemParser.readObject();
                        if (object instanceof X509CertificateHolder x509CertificateHolder) {
                            return new JcaX509CertificateConverter().getCertificate(x509CertificateHolder);
                        }
                    }
                }
            }
        }
        throw new MsranJcatException(String.format("error CA trust certificate pem file!%n{}", certificatePemFilePath));
    }
```

**Comment by hazendaz** [comment]
@mpet Can you try the latest and confirm if issue persists?

**Comment by mpet** [comment]
@hazendaz  I will verify with:

<plugin>
        <groupId>com.github.spotbugs</groupId>
        <artifactId>spotbugs-maven-plugin</artifactId>
         <version>4.8.2.0</version>
      </plugin>

**Comment by mpet** [comment]
When I tested with this version it is ok. Do @hazendaz  or @gtoison know in which version this was fixed?

**Comment by gtoison** [comment]
My understanding (based on https://github.com/spotbugs/spotbugs/pull/1435#issuecomment-783161822 and https://github.com/spotbugs/spotbugs/issues/1992) is that the issue was actually caused by compilers (ecj, javac) producing some redudant bytecode.

Maybe while testing with the gradle plugin you also changed the compiler?

It is not clear to me what was fixed (nor where or when). From what I can see only a unit test reproducing the issue was added to the SpotBugs project, but I might have missed something

**Comment by mpet** [comment]
I am a bit confused now since when I run with the spotbugs maven plugin,spotbugs-maven-plugin, on the below code:
https://github.com/manikmagar/java14-examples/tree/master

Both spotbugs versions  4.7.3.0 and 4.8.2.0

![image](https://github.com/spotbugs/sonar-findbugs/assets/1445826/83c7e8d9-d27f-4957-b9cd-fbc44804f345)

![image](https://github.com/spotbugs/sonar-findbugs/assets/1445826/c10621c2-45bb-42ba-bc3e-c9d756e6dd01)

But I get a different bug reported?!

**Comment by gtoison** [comment]
I think you need to share a way of reproducing the problem or we won't be able to help.
My understanding is that the issue is with the compiler, because some compilers produce redundant bytecode.
Sharing a maven project is not enough because there's no way to tell what compiler was used: even though the above project is java 14 based on the pom, you might be running maven on top of a jdk 21.
Similarly your initial question says you are on java 17 but I'm not reproducing the problem, so I'm probably missing some crucial details.

**Comment by mpet** [comment]
This is really confusing since the messge from the people admin the SQ server was:
"Currently we are using:
FindBugs 4.2.3 / SpotBugs 4.7.3"
is it not only one plugin for both?

**Comment by mpet** [comment]
@gtoison we have been running it on windows and linux but the jdk 17 is the same from Azul.
I got this info from SQ admin.

![image](https://github.com/spotbugs/sonar-findbugs/assets/1445826/2803b69a-d64d-42b3-98a3-d08cee5a1ecf)

**Comment by gtoison** [comment]
I think that you are referring to the JDK running your SonarQube server, but that shouldn't matter here.

**Comment by gtoison** [comment]
There was no feedback so I'm assuming that the issue was with ECJ, please do not hesitate to revert if the problem persists.
